sources/softwareupdate.cpp

```diff
@@ -0,0 +1,239 @@
+            QString     nthree           = QString("%1").arg(newVersionDigits[2].toInt(), 2, 10, QChar('0'));
+
+            int combinedNewVersion;
+            combinedNewVersion = (none.append(ntwo).append(nthree)).toInt();
+
+            QStringList currentVersionDigits = m_currentVersion.split(".");
+            QString     cone                 = QString("%1").arg(currentVersionDigits[0].toInt(), 2, 10, QChar('0'));
+            QString     ctwo                 = QString("%1").arg(currentVersionDigits[1].toInt(), 2, 10, QChar('0'));
+            QString     cthree               = QString("%1").arg(currentVersionDigits[2].toInt(), 2, 10, QChar('0'));
+
+            int combinedCurrentVersion;
+            combinedCurrentVersion = (cone.append(ctwo).append(cthree)).toInt();
+
+            qCDebug(CLASS_LC) << "New version:" << combinedNewVersion << "Current version:" << combinedCurrentVersion;
+
+            // check if it's a newer version than we have
+            if (combinedNewVersion > combinedCurrentVersion) {
+                m_updateAvailable = true;
+                emit updateAvailableChanged();
+
+                // send a notification
+                QObject *param = this;
+                Notifications::getInstance()->add(
+                    false, tr("New software is available"), tr("Download"),
+                    [](QObject *param) {
+                        SoftwareUpdate *i = qobject_cast<SoftwareUpdate *>(param);
+                        i->startUpdate();
+                    },
+                    param);
+            } else {
+                m_updateAvailable = false;
+                emit updateAvailableChanged();
+            }
+        }
+    } else {
+        qCWarning(CLASS_LC) << "Network reply error: " << reply->errorString();
+        Notifications::getInstance()->add(true, tr("Cannot connect to the update server."));
+    }
+
+    reply->deleteLater();
+    qCDebug(CLASS_LC) << "Network reply deleted";
+
+    m_manager->deleteLater();
+    qCDebug(CLASS_LC) << "Networkaccessmanager deleted";
+}
+
+void SoftwareUpdate::downloadUpdate(QUrl url) {
+    qCDebug(CLASS_LC) << "Downloading file" << url.toString();
+
+    qCDebug(CLASS_LC) << "DIR" << QDir(m_appPath).mkpath("downloads");
+
+    QString path = m_appPath.append("/downloads/latest.zip");
+
+    // open the file
+    m_file = new QFile(path);
+    m_file->open(QIODevice::WriteOnly);
+    qCDebug(CLASS_LC) << "File opened";
+
+    // create network manager
+    m_manager = new QNetworkAccessManager(this);
+    QNetworkRequest request(url);
+    request.setAttribute(QNetworkRequest::FollowRedirectsAttribute, true);
+    m_download = m_manager->get(request);
+
+    connect(m_download, &QNetworkReply::downloadProgress, this, &SoftwareUpdate::downloadProgress);
+    connect(m_download, &QNetworkReply::finished, this, &SoftwareUpdate::downloadFinished);
```

Comment:
Error handling is missing if download fails for any reasons!
The QNetworkReply::error(QNetworkReply::NetworkError code) signal must handled as well.